docker/spark-history-server/Dockerfile

```diff
@@ -0,0 +1,96 @@
+#
+# (C) Copyright IBM Corp. 2021, 2021
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# The official ubi8 image from redhat's registry is one of the few accepted base images for Alvearie
+ARG BASE_IMAGE=registry.access.redhat.com/ubi8
+ARG BASE_IMAGE_VERSION=8.4-206.1626828523
+
+ARG SPARK_VERSION=spark-3.1.2
+ARG SPARK_DIST=bin-hadoop3.2
+ARG SPARK_UID=185
+ARG SPARK_WORK_DIR=/spark-work-dir
+ARG SPARK_HOME=/opt/spark
+
+
+#################
+# spark-builder #
+#################
+FROM ${BASE_IMAGE}:${BASE_IMAGE_VERSION} AS spark-builder
+
+ARG SPARK_HOME
+ARG SPARK_VERSION
+ARG SPARK_DIST
+
+# Download and extract Spark
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN curl "https://archive.apache.org/dist/spark/${SPARK_VERSION}/${SPARK_VERSION}-${SPARK_DIST}.tgz" | tar -xz
+# Move the extracted spark dist to SPARK_HOME
+RUN mv "${SPARK_VERSION}-${SPARK_DIST}" "${SPARK_HOME}"
+# Move entrypoint.sh to SPARK_HOME
+RUN mv "${SPARK_HOME}/kubernetes/dockerfiles/spark/entrypoint.sh" "${SPARK_HOME}"
+# Move decom.sh to SPARK_HOME
+RUN mv "${SPARK_HOME}/kubernetes/dockerfiles/spark/decom.sh" "${SPARK_HOME}"
+# Remove the `tini` use from entrypoint.sh.  Tini is unavailable on Red Hat.
+RUN sed -i 's_/usr/bin/tini -s --__g' "${SPARK_HOME}/entrypoint.sh"
+
+
+##############
+# spark-base #
+##############
```

Comment:
Pedantic: What was the reason for splitting the base image into two parts (`spark-builder` and `spark-base`)?

Reply:
This is mainly to still use `RUN` commands for readability, but limit the layers in the final image.

The Docker docs on [multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/#before-multi-stage-builds) goes into a bit more detail:
> ... artificially compresses two RUN commands together using the Bash `&&` operator, to avoid creating an additional layer in the image. This is failure-prone and hard to maintain. It’s easy to insert another command and forget to continue the line using the `\` character...